_posts/2023-01-04-trino-2022-the-rabbit-reflects.md

```diff
@@ -0,0 +1,581 @@
+
+Despite all the hype about vectorization being a silver bullet to make databases
+go fast, the real speed comes from [better algorithms and better data structures
+that lead to lower resource consumption](https://www.youtube.com/clip/UgkxQwDYDS6evVJelNVjWAgrIhzg_Q-cAEyq).
+Here are a few that made their way into Trino this year:
+
+* Trino now offers improved performance for a variety of operations, including
+ complex join criteria pushdown to connectors, faster aggregations, faster
+ joins, and better performance for large clusters. We have also implemented
+ improvements specifically for aggregations with filters and for the Glue
+ metastore. In addition, we now support dynamic filtering for various connectors
+ and have faster query planning for Hive, Delta Lake, Iceberg, MySQL,
+ PostgreSQL, and SQL Server connectors.
+* Along with general performance optimizations, there have been a great deal of
+ query planning optimizations that lead to better performance for specific SQL
+ operators. These include faster `INSERT` queries, improved performance for
+ `LIKE` expressions and highly selective `LIMIT` queries, and enhanced
+ performance and reliability for `INSERT` and `MERGE` operations. We also made
+ performance improvements for `JOIN`, `UNION`, and `GROUP BY` queries, as well
+ as faster planning of queries with `IN` predicates. We have also added support
+ for correlated subqueries in `DELETE` queries.
+* There are also optimizations for specific SQL types’ performance like of
+ string processing, `MAP` and `ROW` types, the `DECIMAL` type, and `ROW` data
+ types. We have also made aggregations over `DECIMAL` columns faster and
+ improved the performance of `ROW` type and aggregation.
+* A last set of improvements come from reading open file formats like ORC and
+ Parquet efficiently. We have improved the speed of reading decimal values in
+ Parquet data, writing `ARRAY`, `MAP`, and `ROW` types to Parquet, and reading
+ Parquet data in general. We have also made writing non-structural types to
+ Parquet faster, providing improved performance for a wide range of
+ Parquet-related operations. ORC data more efficiently. We have made
+ improvements to the performance of ORC in general, and have added the ability
+ to write Bloom filters in ORC files, and implemented these filters for
+ `VARCHAR` columns, further improving the efficiency of ORC-related operations.
+
+These improvements in aggregate are at the core of what makes Trino fast. There
+are rarely silver bullets like vectorization that can improve the overall speed.
+
+### Runtime improvements
+
+The biggest overall runtime improvement comes from the Java ecosystem. Java 17 was the next LTS from Trino’s existing version Java 11.
+
+* Java 17 (390) {{site.url}}/episodes/36.html {% post_url 2022-07-14-trino-updates-to-java-17 %} https://github.com/trinodb/trino/issues/9876
+* Experimental docker image for ppc64le. (393)
+* Add CLI support for ARM64. (370)
+  
+### Security
```

Comment:
Seems like a very long section now .. can we shorten?

Reply:
Removed the last bullet